File: STACK/JavaImplementation/Stack using linkedList/src/com/javalearning/StackUsingLinkedList.java
package com.javalearning;

public class StackUsingLinkedList {

    private class node {
        int data;
        node link;

    }

    node top;

    public StackUsingLinkedList() {
        this.top = null;
    }

    public boolean isEmpty() {
        if(this.top == null) {
            return true;
        } else {
            return false;
        }
    }

    public boolean isFull() {
        node temp = new node();
        if(temp == null) {
            return true;
        } else {
            return false;
        }
    }

    public boolean push(int x) {

        node temp = new node();

        //memory of heap overflow
        if(temp == null) {
            System.out.println("Stack overflow.");
            return false;
        }

        //
        temp.data = x;
        temp.link = top;
        top = temp;
        System.out.println(x + " pushed to stack.");
        return true;
    }

    public int pop() {

        if(isEmpty()) {
            System.out.println("Stack Underflow.");
            return 0;
        } else {
            int x = top.data;
            top = top.link;
            System.out.println(x + " is poped from stack.");
            return x;
        }
    }

    public int peep(int index) {

        int count = index-1;
        node temp = top;

        if(isEmpty()) {
            System.out.println("Stack is empty.");
            return 0;
        }

        while (count != 0 && temp!= null) {
            count--;
            temp = temp.link;
        }

        if(temp == null) {
            System.out.println("Invalid index value.");
            return 0;
        } else {
            System.out.println("Element at position " + index + " from top is " + temp.data);
            return temp.data;
        }
    }

    public int peek() {
        if(isEmpty()) {
            System.out.println("Stack is empty");
            return 0;
        } else {
            System.out.println("Top most element is " + top.data);
            return top.data;
        }
    }

    public void display() {
        node temp = top;
        while ( temp!= null) {
            System.out.print(temp.data + " -> ");
            temp = temp.link;
        }
        System.out.println("Null");
    }
}

File: STACK/stack_using_linkedList.cpp
#include<iostream>
using namespace std;

class Stack {
    class node{
    public:
        int data;
        node *link;
};
    node *top;

public:
    Stack() {
        top = NULL;
    }
    bool isEmpty();
    bool isFull();
    bool push(int x);
    int pop();
    int peep(int index);
    int peek();
    void display();
};

bool Stack :: isEmpty() {
    if(top == NULL) {
        return true;
    } else {
        return false;
    }
}

bool Stack :: isFull() {
    node *temp = new node();
    if(temp == NULL){
        delete temp;
        return true;
    } else {
        delete temp;
        return false;
    }
}

bool Stack :: push(int x) {
    node *temp = new node();
     if(temp == NULL){
        cout<<"Stack overflow."<<endl;
        return false;
    }
    temp->data = x;
    temp->link = top;
    this->top = temp;
    cout<<x<<" pushed to stack."<<endl;
    return true;
}

int Stack :: pop() {
    if(isEmpty()) {
        cout<<"Stack underflow."<<endl;
        return 0;
    }
    node *temp = top;
    int x = temp->data;
    top = top->link;
    delete temp;
    cout<<x<<" poped from stack."<<endl;
    return x;
}

int Stack :: peep(int index) {

    if(isEmpty()) {
        cout<<"Stack is empty"<<endl;
        return 0;
    }
    node *temp = top;
    int count = index-1;
    while(count != 0 && temp!=NULL) {
        temp = temp->link;
        count--;
    }
    if(temp == NULL) {
        cout<<"Invalid index"<<endl;
        return 0;
    } else {
        cout<<"Element at position "<<index<<" from top is "<<temp->data<<endl;
        return temp ->data;
    }
}

int Stack :: peek(){
    cout<<top->data<<" is at top of stack."<<endl;
}

void Stack :: display() {
    node *temp = top;
     while(temp != NULL) {
        cout<<temp->data<<" -> ";
        temp = temp->link;
     }
     cout<<"null"<<endl;
}

int main() {
    Stack st;
    st.push(5);
    st.push(6);
    st.push(9);
    st.display();
    st.peek();
    st.pop();
    st.peep(1);
    st.peep(7);
    st.display();
}
